### PR TITLE
Add CalculationReport tools

### PR DESCRIPTION
**Pull request by SJuliez**
I noticed that I need to write a conversion report for AlphaStrike conversion. Looking at the BV or cost reports, these are really done per pedes with mountains of stringbuilder.append(ROW_START). I wanted a tool for this, so this is the result and I'd rather add it now than keep it buried in the AS conversion branch.

This consists of an interface "CalculationReport" that writes up ths standard interaction with any of the report classes, such as addLine(String, String, String), addHeader() and so on.
The interface is implemented by four classes. These are:
- HTMLCalculationReport: This creates a report much like it's done now, only without having to use any HTML manually
- SwingCalculationReport: This creates a report using Swing Labels, ready to deploy in a panel or wherever; this is a bit superior to HTML as it knows how wide it wants to be and it is easier to format if any formatting is required. (see screen below)
- TextCalculationReport: This is a pure text report, formatted for display in a monospace font. Good for export (see screen below)
- In addition, there is FlexibleCalculationReport, which internally builds all three of the above and has any of them ready, for example for export to clipboard.

Adding a line to any of the four is the same regardless of which Report type is used (and they can be typed as CalculationReport instead of their class): addLine("Damage:", "2+4+6", " = 12"); Adding any of them to a dialog is simple as they all have a toJComponent() method.

For the examples below, I changed the bv calculation of Mech to the new way (which cuts around 700 lines of code and, I believe, a few bugs along the way). This is not part of the PR as I would need to convert all unit types and thats a lot. Something for later.
The code using it starts with:
bvReport.addHeader("Battle Value Calculations For " + getChassis() + " " + getModel());
bvReport.addSubHeader("Defensive Battle Rating Calculation:");

To switch between these two screenshots I changed only one line:
CalculationReport bvReport = new TextCalculationReport(); <-> CalculationReport bvReport = new SwingCalculationReport();
(and set the font to monospaced for the text)
image
image

(There is a bug in the screenshots that has been corrected in the code)

### CODE REVIEW

File: megamek/src/megamek/client/ui/swing/calculationReport/CalculationReport.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+     *
+     * @param type The first element of this line, such as "Damage: "
+     * @param calculation A calculation or other info, displayed after the type
+     * @param result A result or other info, displayed on the right side
+     * @return The CalculationReport itself. Enables multiple stringed addLine calls
+     */
+    CalculationReport addResultLine(@Nullable String type, @Nullable String calculation, @Nullable String result);
+
+    /**
+     * Returns the CalucaltionReport as a JComponent that can be added to a dialog or other Swing component.
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/calculationReport/CalculationReport.java:191]:
```suggestion
     * Returns the CalculationReport as a JComponent that can be added to a dialog or other Swing component.
```

File: megamek/src/megamek/client/ui/swing/calculationReport/HTMLCalculationReport.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+    @Override
+    public JComponent toJComponent() {
+        final JEditorPane editorPane = new JEditorPane("text/html", this.toString());
+        editorPane.setEditable(false);
+        editorPane.setCaretPosition(0);
+        return editorPane;
+    }
+
+    public CalculationReport addResultLine(String type, String calculation, String result) {
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/calculationReport/HTMLCalculationReport.java:58]: Missing override annotation

File: megamek/src/megamek/client/ui/swing/calculationReport/SwingCalculationReport.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    private final GridBagConstraints gbc = new GridBagConstraints();
+    private final String COL_SPACER = "   ";
+    private final static int PADDING = 15;
+
+    @Override
+    public JComponent toJComponent() {
+        return report;
+    }
+
+    public CalculationReport addResultLine(String type, String calculation, String result) {
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/calculationReport/SwingCalculationReport.java:39]: Missing override annotation

**Review by Windchild292 — COMMENTED** [review]
Looks good, just a typo and a few missing overrides

**Comment by SJuliez** [comment]
OK corrected.

**Review by Windchild292 — APPROVED** [review]

**Pull request merged**